Refactoring the feature system to remove the `cast_feature_ref` function

Janonard opened the pull request:
Okay, this is a bit weird because I'm reverting some of your changes, but I have good reasons to do so:

At first, I didn't like the `cast_feature_ref` function because it is written without a context, like a type or trait. But secondly and more importantly, it creates a dummy reference to a generally unknown but zero-sized type if the provided pointer points to null. This is very wrong since the reference isn't actually of the given type and therefore, it technically isn't valid.

The first thing could easily be fixed: Move that function inside the `Feature` trait and you're done. The second thing didn't went so easy, because it's roots lie in a design decision: The feature container always has to contain references to the features and this simply doesn't work: Some feature's data pointer does not point to null because it's just a flag, it points to null because there simply is no data to reference. Creating a dummy feature reference in this context is not the right thing to do and also removes freedom from the feature implementer: What if the host does not provide data, but the implementer wants to store it's own data anyway? This would simply not work with this solution.

Therefore, I needed to refactor: Now, every feature has to implement a method that tries to construct the feature struct instance from a data pointer. If the data pointer is used, it has to be stored inside the feature struct as a reference, if not, not. All semantics of the retrieval are contained in this method, which provides maximum flexibility. The derive macro is capable of handling features with and without lifetimes, even complete feature collections without lifetimes. 

On the plugin's side of things, this went relatively easy, but when I started to port the URID module, I ran into problems regarding the host code: It very much depended on the raw features being more or less the same as the wrapper features, which already felt a bit shaky and now didn't work anymore. I needed to rework that too.

This also gave me the opportunity to introduce pins to the system: Since we store a lot of raw pointers to things, the compiler does not assure that they are always valid. Therefore, we have to do that on our own by pinning the pointees. First, we create the mapper and then create the interfaces. These interfaces contain a clone of the mapper, the raw feature, and the feature descriptor, all boxed and pinned. By the way, the map and unmap interfaces are split because having two structs containing mutable pointers to the same value is against the spirit, maybe even against the rules of Rust. The only way the internal pointers may become invalid is when the boxes are dropped, which we can't influence and would be done by the actual host.

The overall process is maybe still a bit rough and could be smoothed out, but it points in a good direction on how to handle host code. However, this would be a topic for another PR.

prokopyl commented:
> Okay, this is a bit weird because I'm reverting some of your changes, but I have good reasons to do so:

Yes, I feel this is getting really awkward as well. If there are any issues with the current (accepted) design they should be discussed before submitting new code, I feel we are throwing a lot of code through the window lately and this may become quite frustrating for both of us. :/

> At first, I didn't like the `cast_feature_ref` function because it is written without a context, like a type or trait.

I have noticed you do not seem to like free-standing functions in general, why is that? Functions do not have to be necessarily tied to anything if it does not make sense for them to be (see for instance the functions in [`serde_json`](https://docs.rs/serde_json/1.0.40/serde_json/), or the many private utility functions that are scattered around the implementation `std` library).

> The first thing could easily be fixed: Move that function inside the `Feature` trait and you're done. 

Also, putting it back onto the `Feature` trait is probably the worst place to put it in. While putting it on a type vs. making it free-standing is purely a question of semantics ("does that function work on data from the given type?" is how I generally think about that issue), putting it on a trait should not be done only for code organization reasons, because it is a completely different contract with the user!

All methods present on a trait bring the following contract with the users: if you implement the trait, you have to implement the method, or if it has an implementation already you can override it. The method is also publicly accessible on all types that implement it.

For instance calling `urid::Map::cast_feature_ref()` does not make any sense, and it doesn't make any more sense for any feature to override it either.
Plus (and this should be the main argument after all), this function is internal (it's private, only accessible to the module) and for internal use only, and it should stay that way. Exposing it to the user is not a good idea, and polluting every single feature with it is even worse, I think.

Of course the right fix here is not to hide this function, but to simply remove it altogether from the code (and by that I mean killing it with fire). :)

I know you removed the function already, but this sentence of yours triggered me about a more general context: functions that are internal should stay so, even if that means we have to put them "aside" (which really isn't a bad thing in general anyway).

> But secondly and more importantly, it creates a dummy reference to a generally unknown but zero-sized type if the provided pointer points to null. This is very wrong since the reference isn't actually of the given type and therefore, it technically isn't valid.

That I absolutely agree with. It works in practice because the type is zero-sized, but the simple fact that I had to check the disassembly and make extra tests to be sure it does work (and the fact that there are 3 paragraphs of not really documentation but more like justification above it) shows it is really way too much of an hack for me to be comfortable keeping this in code anyway. :) 

> The second thing didn't went so easy, because it's roots lie in a design decision: The feature container always has to contain references to the features

Actually, this isn't true anymore: with the adjustments I made to handle `Option`s , any type can be used in a feature container now, as long as it implements the right trait.

Maybe we could just use it a bit more so we make "Marker" features (the zero-sized ones) to be only put in there by value instead of by reference?

> Creating a dummy feature reference in this context is not the right thing to do and also removes freedom from the feature implementer: What if the host does not provide data, but the implementer wants to store it's own data anyway? This would simply not work with this solution.

This "freedom" should not be needed, because the issue of storing additional data is completely orthogonal to the use of features. A `Feature` should map exactly to an LV2 feature and be completely stateless, without any additional data on the client/plugin side, only encapsulated in a safe interface for the Rust side. The difference is here: with your idea a `Feature` contains/uses an LV2 feature, while I think a `Feature` *is* just the feature, without any extra bloat.

This is because if the implementer wants to store it's own data, then that data is, by definition, not the feature's, which is only what is defined by the spec. What you call "implementer's data" is not really data, it's additional state.

Now, if you want to provide something that *extends* a feature by using it and has its own extra state, you can already do so with simple composition, no need for all that extra fluff!

Here is an example with the idea of caching calls to `urid::Map` locally you had at some point (this is just to illustrate that it is possible, not really a good idea in practice.):

```rust
struct URIDCacheMap<'a> {
  map_feature: &'a ::urid::feature::Map,
  cache: HashMap<String, URID>
}

impl<'a> URIDCacheMap<'a> {
  fn new(map_feature: &'a ::urid::feature::Map) -> Self {
   // ...
 }
}
```

And done! That way the user can also choose what to use simply by wrapping it or not.
You couldn't (and shouldn't) put that struct into the feature collection however, because it can be both shared and used simultaneously across threads (see the worker or state specs for instance). Which is also another reason why features should be stateless client-side. :)

> This also gave me the opportunity to introduce pins to the system: Since we store a lot of raw pointers to things, the compiler does not assure that they are always valid. Therefore, we have to do that on our own by pinning the pointees.

I'm not sure what the purpose of pins are in here: wasn't the lifetime on the `Map` feature struct enough to prevent both moving, mutating and dropping, without forcing extra allocation? E.g:

```rust
use std::marker::PhantomData;

struct Mapper { /* */ }
struct MapFeature<'a> { mapper: *const Mapper, _lifetime: PhantomData<&'a Mapper> }

impl<'a> MapFeature<'a> {
    pub fn new(mapper: &'a Mapper) -> Self {
        Self { mapper: mapper as *const _, _lifetime: PhantomData }
    }
    
    pub fn foo(&self) {}
}

fn main() {
    let mapper = Mapper {};
    let feature = MapFeature::new(&mapper);
    move_mapper(mapper); // Compiler says: Nope
    feature.foo();
}

fn move_mapper(_m: Mapper) {}
```

To me this seemed to be working well, why replace it?

> By the way, the map and unmap interfaces are split because having two structs containing mutable pointers to the same value is against the spirit, maybe even against the rules of Rust. 

I'm not sure what you are talking about here, all references to and from `Mapper`s were shared, immutable references, so there is no issue here.

Unless you are talking about the pointers defined in spec (both in core and URID), where all pointers are `*mut`, in which case it would be, but remember that pointers are not references! In both C and unsafe Rust, `*const T` and `*mut T` have very little difference, because there is no concept of immutability in C (since you can cast that away very easily). In rust however, `&T` and `&mut T` respectively mean "shared pointer" and "unique pointer". And in LV2, all features are referenced multiple times (at least once per live plugin), so they are shared (`&T`) pointers, but this restriction only applies in Rust, C can definitely mutate objects through shared pointers.

You should not assume `*mut` means `&mut` in C function definitions, only assume uniqueness/mutability based on what the spec or docs say, i.e. something like "calling this invalidates all references to the object", which clearly indicates uniqueness.

Also, I really do not like the concept of "interfaces" (which is applied here on `Map`/`Unmap`, but it will be the same with every single feature that needs host data if I follow your logic). They add a lot of bloat, both externally (2 new structs, that are clearly foreign to the LV2 spec, that the user has to instantiate to get their features), and internally (since all values are `Box`ed, which adds quite a bit of extra, unnecessary allocation).

To me, holding simple `&T` with features having lifetimes works well enough without bringing all of this extra complexity, while enforcing everything is safe and sound in Rust <=> Rust LV2 communication.

> The only way the internal pointers may become invalid is when the boxes are dropped, which we can't influence and would be done by the actual host.

Mappers and all are host-only, so yes we can influence it, because we are making the interface to it, if there are any soundness holes then we must fix them. The current solution (with lifetimes) works well enough and does not have any of these soundness issues, either.

prokopyl commented on core/src/feature/container.rs:
```diff
@@ -78,27 +77,27 @@ impl<'a> FeatureCollection<'a> for FeatureContainer<'a> {
 }
 
-/// A trait to allow arbitrary types to be potentially created from feature resolution.
+/// A trait to allow arbitrary types to be potentially created from a feature resolution.
```
Nitpick here, but I meant "feature resolution" as a general process, there is no "Feature resolution" object so having "a" in front of it is incorrect, I believe. :slightly_smiling_face:

prokopyl commented on urid/src/mapper.rs:
```diff
@@ -53,12 +114,50 @@ pub trait URIDMapper {
-#[derive(Default)]
-pub struct HashURIDMapper(Mutex<HashMap<UriBuf, URID>>);
+#[derive(Default, Clone)]
+pub struct HashURIDMapper(Arc<Mutex<HashMap<UriBuf, URID>>>);
```
Having a struct containing an Arc around its data and then implement clone on it is really unexpected behavior for users and should be avoided: when calling `clone()` on an object users expect it to be a deep clone, but here it makes just a simple reference count increment and still shares the same data.

`Rc` and `Arc` here should be created/used by the user explicitly, and only if needed (you can use `Borrow` if you want to be generic over it), see the [`Arc` docs](https://doc.rust-lang.org/std/sync/struct.Arc.html#cloning-references) and the associated [Clippy lint](https://rust-lang.github.io/rust-clippy/master/#clone_on_ref_ptr).

prokopyl commented on urid/src/mapper.rs:
```diff
@@ -40,6 +64,43 @@ pub trait URIDMapper {
     fn map(&self, uri: &Uri) -> Option<URID>;
 
+    /// Unsafe wrapper of the `map` method, used by the feature interface.
+    unsafe extern "C" fn extern_map(
```
Remember putting a function on a trait makes it public, callable and override-able! You have to put this method separate from the trait! (See my comment on the PR)

Janonard replied:
I actually see no problem with that, as stated before!

Janonard commented:
> I have noticed you do not seem to like free-standing functions in general, why is that? Functions do not have to be necessarily tied to anything if it does not make sense for them to be (see for instance the functions in serde_json, or the many private utility functions that are scattered around the implementation std library).
> [...]
> While putting it on a type vs. making it free-standing is purely a question of semantics ("does that function work on data from the given type?" is how I generally think about that issue)

When a function is bound to a type, it means to me that this method "belongs" to that type, it does not necessarily have to work on data of that type. I agree with you that functions may be free-standing, but only if you can say they don't "belong" to any type. In this case, I felt that this function "belongs" to `Feature`.

> All methods present on a trait bring the following contract with the users: if you implement the trait, you have to implement the method, or if it has an implementation already you can override it.

Ehm, no. For example, take a look at the [Iterator trait](https://doc.rust-lang.org/std/iter/trait.Iterator.html). It has only one method that should be implemented by the user, but docents of predefined methods which make iterators really handy. If you want to, you can override them, but you shouldn't or at least don't need to.

Just because a method is part of a trait, it does not mean that you should override it; It just means that it "belongs" to that trait; That every type that implements the trait has these methods. The external map and unmap methods mimic that style: They enhance the trait by building upon other, user-defined methods. The user does not have to override them, but if they wish so, they could; It's not our business. This style is canonical; I see no reason not to use it.

> Maybe we could just use it a bit more so we make "Marker" features (the zero-sized ones) to be only put in there by value instead of by reference?

This is what I effectively did! :smile: 

> This "freedom" should not be needed, because the issue of storing additional data is completely orthogonal to the use of features.

Jepp, you're right about that. It is a bad example, but it does not change the fact that the dummy reference is not the best solution, is it? :wink: 

> The difference is here: with your idea a `Feature` contains/uses an LV2 feature, while I think a `Feature` *is* just the feature, without any extra bloat.

Well, you already said it: I think that a `Feature` should contain/use a `sys` feature struct and not try to act as if it were the same, because it is not. Currently, there are always two structs: The one generated from the C header and the one that implements the `Feature` trait. The generated struct is instantiated by the host, which passes a pointer to that instance to the plugin. But then, instead of simply accepting that instance is what it is, the plugin acts as if the instance is of a different, plugin-defined type. Why does it have to do that if it could simply work on a reference to the `sys` feature struct?

Also, accepting that these structs aren't the same doesn't add any bloat: The `Map` and `Unmap` structs transparently contain a reference to their respective `sys` structs, which they received from the host. This means that in memory, they are just references, nothing more, and their methods simply work on references too. There is no bloat.

> I'm not sure what the purpose of pins are in here: wasn't the lifetime on the Map feature struct enough to prevent both moving, mutating and dropping, without forcing extra allocation? E.g:

Yes, this helps inside the map feature struct, but not, for example, in `sys::LV2_URID_Map`: The pointer to the mapping handle is not guarded by any lifetimes and we therefore have to assure that it stays valid when we construct it on our own. This means that the handle, in our case the struct that implements `URIDMapper`, has to be pinned in memory. Also, I see no problem in allocating it dynamically: No proper host will do that in the real-time thread and we can therefore allocate as much as we want! :smile: 

However, we should use those phantom references wherever it is possible, because they also prevent the referenced data from being dropped. I definitely will reintroduce them!

> I'm not sure what you are talking about here, all references to and from Mappers were shared, immutable references, so there is no issue here.

I'm talking about `sys::LV2_URID_Map` again: It contains a mutable pointer to a handle. Although it is not against the rules of Rust to have multiple mutable pointers to one struct instance, you could dereference them "by accident". Then, you would have multiple mutable references to one struct instance, which is very much against the rules indeed! Therefore, I gave every `sys::LV2_URID_Map` it's own smart pointer instance to point to. If you would dereference all map handler pointers at once now, you would always have exactly one mutable reference to every smart pointer.

> Also, I really do not like the concept of "interfaces" (which is applied here on Map/Unmap, but it will be the same with every single feature that needs host data if I follow your logic).

I'm not super-happy with them either, I will remove them if you wish so.

Janonard commented:
I've removed most of the pinned boxed now. However, the `make_X_interface` methods still require the mutable references to the mappers to be pinned, due to the discussed reasons. I've also removed the `Arc`s, which would technically enable the scenario with multiple mutable reference to the mapper, but made the transition a bit easier. Maybe I will reintroduce some kind of smart pointer later.

Janonard commented:
> To me this seemed to be working well, why replace it?

I just noticed: We were talking about two different things: You are talking about `lv2_urid::feature::Map` and that it contains a reference to a raw feature which prevents the raw feature from dropping. What you said about it is absolutely correct and not a subject of change, even better: It stores a proper reference to the raw feature, not just a phantom!

I was talking about the host side of things, the structs `lv2_urid_sys::LV2_URID_Map` and `lv2_core_sys::LV2_Feature`: We can not modify them and therefore we have to assure the validity of their internal pointers from the outside. This is done with pins.

Janonard commented:
@prokopyl There is only one open point in your review, which I explained in great detail. What is your opinion on that now?
